### PR TITLE
Fix printing of functor types

### PR DESCRIPTION
**Pull request by 314eter**
The module type `(S -> S) -> S` is currently printed as `S -> S -> S`, which is equal to `S -> (S -> S)`.

### CODE REVIEW

File: parsing/pprintast.ml
Comment (trefis) [inline parsing/pprintast.ml]: Can you explain why you add parentheses around a functor only when there are no attributes, aren't they also necessary if there are some attributes?

Comment (gasche) [reply parsing/pprintast.ml]: This is standard in `foo1` functions (see for example [core_type1](https://github.com/314eter/ocaml/blob/9124ce56f95e2480a50062f8897fc80a684f14fc/parsing/pprintast.ml#L273-L274), to defer to the toplevel `foo` function in case of attributes, because it prints the attributes with a lot of parentheses for safety -- see [attribute printing in module_type](https://github.com/314eter/ocaml/blob/9124ce56f95e2480a50062f8897fc80a684f14fc/parsing/pprintast.ml#L957-L959).

On the other hand, the structure of `module_type1` is non-standard: these higher-level printers should handle the constructions that they know do *not* need parethesizing, and defer the others to higher-level printers *or* add parentheses to go back to the toplevel printer (`module_type` here). This is not what this function does, and I am not sure it is correct (for example, it would be wrong if given `(S1 -> S2) with eqn` as argument; but this construction is not type-correct).

I would expect the following two levels:
- `simple_module_type` which prints all constructions that are fully parenthesized (`sig..end`) or atomic (identifiers), for others it calls `module_type` and wraps parentheses around it
- `module_type` which prints the rest, and in particular the `Pmty_functor` construction; in the `S1 -> S2` case, it calls `simple_module_type` on the left-hand-side.
Maybe it would be more correct

**Review by gasche — CHANGES_REQUESTED** [review]

File: parsing/pprintast.ml
Comment (trefis) [reply parsing/pprintast.ml]: Thanks, I had missed that.

Comment (314eter) [reply parsing/pprintast.ml]: I'm using the standard structure for `module_type` and `module_type1` now. I think the previous approach was correct too, but maybe this is more future proof.

File: parsing/pprintast.ml
Comment (gasche) [inline parsing/pprintast.ml]: It is strange that you match the same construction in both functions, maybe remove this `Pmty_with (mt, [])` clause? (It is not wrong, and in theory it could save parentheses in some cases (although I don't think they can actually happen), but I like sticking to the standard pattern.)

File: parsing/pprintast.ml
Comment (gasche) [inline parsing/pprintast.ml:1078]: Why are you using the inner-level here and below? (And not in `Psig_module`.)

File: testsuite/tests/parsetree/source.ml
Comment (gasche) [inline testsuite/tests/parsetree/source.ml:163]: The parentheses here are a consequence of using `type1` above, but why are they desirable/useful?

**Review by gasche — COMMENTED** [review]
I like the new structure better, thanks.

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
(I made a minor mistake in "approving" the changes, I just wanted to dismiss my previous change request which has been addressed. There are minor issues still to discuss, and I think you should have a Changes entry.)

File: parsing/pprintast.ml
Comment (314eter) [reply parsing/pprintast.ml]: I did it to save parentheses, but `Pmty_with (mt, [])` can only occur in artificial parsetrees, so I'll remove it.

File: parsing/pprintast.ml
Comment (314eter) [reply parsing/pprintast.ml:1078]: Otherwise you would get
```ocaml
module rec A : S with type t = t and B : S with type t = t
```
which can't be parsed (the `and B : S` is considered part of the constraint). In principle, it's not really necessary below, but I did it for symmetry.

In `Psig_module` there is no `and` possible. Is it better to be consequent and add parentheses there too?

**Comment by damiendoligez** [comment]
@gasche you could dismiss your review, right?

**Comment by gasche** [comment]
Well yes, but actually @314eter responded to my questions and I am now happy to fully approve it. I may merge soon unless someone objects.

**Comment by damiendoligez** [comment]
ping @gasche

**Comment by gasche** [comment]
I just rebased the branch on top of the current trunk, and will push here when I'm done testing locally, then merge if the CI passes.

**Comment by gasche** [comment]
(This is now rebased and should be ready to merge. Feel free to ping me if you come here and the CI has finished. I have also rebased on 4.07 and plan to push both at once.)

**Pull request merged**
